internal/venice-client-common/src/main/java/com/linkedin/venice/schema/writecompute/WriteComputeSchemaConverter.java

```diff
@@ -318,91 +319,24 @@ private Schema convertMap(Schema mapSchema, String name, String namespace) {
-   *
-   * N.B.: if it's an top level union field, the convert will try to convert the elements(except for RECORD)
-   * inside the union. It's not supported if an union contains more than 1 collections since it will
-   * confuse the interpreter about which elements the operation is supposed to be applied.
-   *
-   * e.g.
-   * original schema:
-   *{
-   *   "type" : "record",
-   *   "name" : "record",
-   *   "fields" : [ {
-   *     "name" : "nullableArrayField",
-   *     "type" : [ "null", {
-   *       "type" : "array",
-   *       "items" : {
-   *         "type" : "record",
-   *         "name" : "simpleRecord",
-   *         "fields" : [ {
-   *           "name" : "intField",
-   *           "type" : "int",
-   *           "default" : 0
-   *         } ]
-   *       }
-   *     } ],
-   *     "default" : null
-   *   } ]
-   * }
-   *
-   * write compute schema:
-   * {
-   *   "type" : "record",
-   *   "name" : "testRecordWriteOpRecord",
-   *   "fields" : [ {
-   *     "name" : "nullableArrayField",
-   *     "type" : [ {
-   *       "type" : "record",
-   *       "name" : "NoOp",
-   *       "fields" : [ ]
-   *     }, "null", {
-   *       "type" : "record",
-   *       "name" : "nullableArrayFieldListOps",
-   *       "fields" : [ {
-   *         "name" : "setUnion",
-   *         "type" : {
-   *           "type" : "array",
-   *           "items" : {
-   *             "type" : "record",
-   *             "name" : "simpleRecord",
-   *             "fields" : [ {
-   *               "name" : "intField",
-   *               "type" : "int",
-   *               "default" : 0
-   *             } ]
-   *           }
-   *         },
-   *         "default" : [ ]
-   *       }, {
-   *         "name" : "setDiff",
-   *         "type" : {
-   *           "type" : "array",
-   *           "items" : "simpleRecord"
-   *         },
-   *         "default" : [ ]
-   *       } ]
-   *     }, {
-   *       "type" : "array",
-   *       "items" : "simpleRecord"
-   *     } ],
-   *     "default" : { }
-   *   } ]
-   * }
+   * Wrap up a union schema with possible partial update operations.
+   * Only two branches union with one being NULL type is allowed. This will make partial update logic compatible with
+   * Active/Active replication schema requirement.
    */
   private Schema convertUnion(Schema unionSchema, String name, String namespace) {
     if (unionSchema.getType() != UNION) {
       throw new VeniceException("Expect schema to be UNION type. Got: " + unionSchema);
     }
-    SchemaUtils.containsOnlyOneCollection(unionSchema);
-    return SchemaUtils.createFlattenedUnionSchema(unionSchema.getTypes().stream().sequential().map(schema -> {
-      Schema.Type type = schema.getType();
-      if (type == RECORD) {
-        return schema;
-      }
-      return convert(schema, name, namespace);
-    }).collect(Collectors.toList()));
+    if (isCollectionUnionSchema(unionSchema)) {
```

Comment:
I have a question about this requirement.
For union with multiple types (1. not nullable union, 2. contains more than 2 types), for write compute, I assume we can still allow the full field update, right? And I assume for RMD schema generation, we can generate a field-level timestamp for such union field, right?

Reply:
What you said might make sense, let me think about it.

Reply:
Yeah, technically yes. 
I don't have a context about why these are not supported in A/A though, I suspect there might be some earlier discussion for A/A implementation that leads to the decision to block it (maybe for simplicity in implementation?)
So in theory, there are all these types:
    SCHEMA TYPE        CURRENT RMD TYPE
1. primitive - long
2. collection - [long, complex record]
3. [primitive] - long (wrong, should be [long, complex record])
4. [collection] - long (wrong, should be [long, complex record])
5. [null, primitive] - long
6. [null, collection] - [long, complex record]
7. [primitive1, primitive2] - long
8. [primitive, collection] - Not supported
9. [collection1, collection2] - Not supported

Union with 3 branches must have same judgement call as type 5-9.
For today, all of them are supported by WC converter, except 9. 

Are you suggesting we should support all these, except 9 which leads to ambiguious RMD data for value?
My current understanding is, technically we can support that, but I feel like in real use case people usually have 2 branches union, and for all Fedex-backed schema, it is always [null, something]...

Reply:
Yeah, most used usecase for union is nullable union. If someone designs [collection1,collection2] as union, they should  redesign the schema and rejecting such schemas for WC should be fine.